library/init/data/nat/bitwise.lean

```diff
@@ -23,30 +23,187 @@ namespace nat
+  | is_false h := (mod_two_eq_zero_or_one _).resolve_right h
+  end
+
+  lemma bit_decomp (n : nat) : bit (bodd n) (shiftr n 1) = n :=
+  (bit_val _ _).trans $ (add_comm _ _).trans $
+  eq.trans (by rw mod_two_of_bodd; refl) (mod_add_div n 2)
+
+  lemma bit_cases_on {C : nat → Sort u} (n) (h : ∀ b n, C (bit b n)) : C n :=
+  by rw -bit_decomp n; apply h
+
+  lemma bodd_bit (b n) : bodd (bit b n) = b :=
+  begin
+    rw bit_val, dsimp [bodd],
+    rw [add_comm, add_mul_mod_self_left, mod_eq_of_lt];
+    cases b; exact dec_trivial
+  end
+
+  lemma shiftr1_bit (b n) : shiftr (bit b n) 1 = n :=
+  begin
+    rw bit_val, dsimp [shiftr],
+    rw [add_comm, add_mul_div_left, div_eq_of_lt, zero_add];
+    cases b; exact dec_trivial
+  end
+
+  def shiftl_add (m n) : ∀ k, shiftl m (n + k) = shiftl (shiftl m n) k
+  | 0     := rfl
+  | (k+1) := congr_arg ((*) 2) (shiftl_add k)
+
+  def shiftr_add (m n) : ∀ k, shiftr m (n + k) = shiftr (shiftr m n) k
+  | 0     := rfl
+  | (k+1) := congr_arg (/ 2) (shiftr_add k)
+
+  def shiftl_eq_mul_pow (m) : ∀ n, shiftl m n = m * 2 ^ n
+  | 0     := (mul_one _).symm
+  | (k+1) := (congr_arg ((*) 2) (shiftl_eq_mul_pow k)).trans $ by simp [pow_succ]
+
+  def one_shiftl (n) : shiftl 1 n = 2 ^ n :=
+  (shiftl_eq_mul_pow _ _).trans (one_mul _)
+
+  def zero_shiftl (n) : shiftl 0 n = 0 :=
+  (shiftl_eq_mul_pow _ _).trans (zero_mul _)
+
+  def shiftr_eq_div_pow (m) : ∀ n, shiftr m n = m / 2 ^ n
+  | 0     := (nat.div_one _).symm
+  | (k+1) := (congr_arg (/ 2) (shiftr_eq_div_pow k)).trans $
+             by dsimp; rw [nat.div_div_eq_div_mul]; refl
+
+  def zero_shiftr (n) : shiftr 0 n = 0 :=
+  (shiftr_eq_div_pow _ _).trans (nat.zero_div _)
+
+  def test_bit_zero (b n) : test_bit (bit b n) 0 = b := bodd_bit _ _
+
+  def test_bit_succ (m b n) : test_bit (bit b n) (succ m) = test_bit n m :=
+  have bodd (shiftr (shiftr (bit b n) 1) m) = bodd (shiftr n m), by rw shiftr1_bit,
+  by rw [-shiftr_add, add_comm] at this; exact this
+
+  def binary_rec {C : nat → Sort u} (f : ∀ b n, C n → C (bit b n)) (z : C 0) : Π n, C n
+  | n := if n0 : n = 0 then by rw n0; exact z else let n' := shiftr n 1 in
     have n' < n, from (div_lt_iff_lt_mul _ _ dec_trivial).2 $
     by note := nat.mul_lt_mul_of_pos_left (dec_trivial : 1 < 2)
          (lt_of_le_of_ne (zero_le _) (ne.symm n0));
        rwa mul_one at this,
+    by rw [-show bit (bodd n) n' = n, from bit_decomp n]; exact 
     f (bodd n) n' (binary_rec n')
 
   def size : ℕ → ℕ := binary_rec (λ_ _, succ) 0
 
   def bits : ℕ → list bool := binary_rec (λb _ IH, b :: IH) []
 
-  def bit (b : bool) : ℕ → ℕ := cond b bit1 bit0
-
   def bitwise (f : bool → bool → bool) : ℕ → ℕ → ℕ :=
   binary_rec
     (λa m Ia, binary_rec
       (λb n _, bit (f a b) (Ia n))
       (cond (f tt ff) (bit a m) 0))
-    (λb, cond (f ff tt) b 0)
+    (λn, cond (f ff tt) n 0)
 
   def lor   : ℕ → ℕ → ℕ := bitwise bor
   def land  : ℕ → ℕ → ℕ := bitwise band
   def ldiff : ℕ → ℕ → ℕ := bitwise (λ a b, a && bnot b)
```

Comment:
I recommend against `|` because this is used in pattern matching and so on - at least I assume that's why subtype changed to `{x // p x}` from lean 2 `{x | p x}`.